pgxn/neon/pagestore_smgr.c

```diff
@@ -1839,6 +1858,14 @@ neon_read_at_lsn(RelFileNode rnode, ForkNumber forkNum, BlockNumber blkno,
 		.blockNum = blkno,
 	};
 
+	/*
+	 * It is possible that we need to read a page through the neon SMgr
+	 * in the backend that is replaying the WAL. In that case, we can't
+	 * wait for that page to be replayed.
+	 */
+	if (RecoveryInProgress())
+		XLogWaitForReplayOf(request_lsn);
+
```

Comment:
I don't understand this. When do you need to wait for some LSN to be replayed? What would go wrong if you just removed this?

Reply:
Ok, understood. I think the comment needs to explain this scenario better. Maybe with an example.

Since this is expected to happen very infrequently, we could perhaps poll GetXLogReplayRecPtr(), instead of adding the condition variable. That way we wouldn't need to make any backend changes. But I think this is fine too.

It might also be useful to add a SQL function to wait for replay to reach a particular LSN. For example, if you have a primary - replica pair, and you make a change in the primary and then want to run a query in the replica so that you see your own changes. It could use the same XLogWaitForReplayOf function.